test/hotspot/gtest/nmt/test_nmt_startup.cpp

```diff
@@ -0,0 +1,224 @@
+  avg /= LOOPS_BENCHMARK;
+  fprintf(stderr, "\nAvg Total Time: %ld us\n", avg/1000);
+
+  jlong allowed = (jlong)(1.25*avg);
+  int counted = 0;
+  avg = 0;
+  for (int t=0; t<LOOPS_BENCHMARK; t++) {
+    if (totals[t] < allowed) {
+      counted++;
+      avg += totals[t];
+    }
+  }
+  avg /= counted;
+  fprintf(stderr, "Clean Avg Total Time: %ld us\n", avg/1000);
+}
+
+static void collect_indexes(int count, int* indexes, void** pointers, DataStruct* data) {
+  for (int i=count-1; i>=0; i--) {
+    pointers[i] = nullptr;
+    indexes[i] = -1;
+    if (data[i].requested > 0) {
+     if (data[i].pointer_prev != 0L) { // realloc
+       indexes[i] = -2; // assume that we will not find the original pointer this realloc references
+       for (int j=i-1; j>=0; j--) {
+         if (data[i].pointer_prev == data[j].pointer) {
+           indexes[i] = j;
+           break;
+         }
+       }
+     }
+   } else if (data[i].requested == 0) { // free
+      indexes[i] = -2; // assume that we will not find the original pointer this free references
+      for (int j=i-1; j>=0; j--) {
+        if (data[i].pointer == data[j].pointer) {
+          indexes[i] = j;
+          break;
+        }
+      }
+    }
+  }
+  for (int i=count-1; i>=0; i--) {
+    if (indexes[i] == -2) {
+      fprintf(stderr, "pointer NOT FOUND\n");
+    }
+  }
+}
+
+void test(int count, int* indexes, void** pointers, DataStruct* data) {
+  // look for pointers which we will need to use later for realloc/free now,
+  // so that we don't include this phase in the performance timing
+  collect_indexes(count, indexes, pointers, data);
+
+  // give the VM time to "settle down"
+  sleep(1);
+
+  fprintf(stderr, "\n\n");
+  run_benchmarks(count, indexes, pointers, data);
+  fprintf(stderr, "\n\n");
+  //MemTracker::tuning_statistics(tty);
+  //MemTracker::final_report(tty);
+}
+
+#if defined(__APPLE__)
+
+TEST_VM(NMTPerformance, test_startup_memory_mac_data) {
+  int count = DATA_MAC_COUNT;
+  ALLOW_C_FUNCTION(::calloc, int* indexes = (int*)calloc(count, sizeof(int));)
+  ALLOW_C_FUNCTION(::calloc, void** pointers = (void**)calloc(count, sizeof(void*));)
+  DataStruct* data = data_mac;
+
+  assert(indexes!=nullptr, "indexes!=nullptr");
+  assert(pointers!=nullptr, "pointers!=nullptr");
+
+  test(count, indexes, pointers, data);
+
+  ALLOW_C_FUNCTION(::free, ::free(pointers);)
+  ALLOW_C_FUNCTION(::free, ::free(indexes);)
+}
+
+#elif defined(LINUX)
+
+TEST_VM(NMTPerformance, test_startup_memory_linux_data) {
+  int count = DATA_LINUX_COUNT;
+  ALLOW_C_FUNCTION(::calloc, int* indexes = (int*)calloc(count, sizeof(int));)
+  ALLOW_C_FUNCTION(::calloc, void** pointers = (void**)calloc(count, sizeof(void*));)
+  DataStruct* data = data_linux;
+
+  assert(indexes!=nullptr, "indexes!=nullptr");
+  assert(pointers!=nullptr, "pointers!=nullptr");
+
+  test(count, indexes, pointers, data);
+
+  ALLOW_C_FUNCTION(::free, ::free(pointers);)
+  ALLOW_C_FUNCTION(::free, ::free(indexes);)
+}
+
+#endif
```

Comment:
What is your plan for Windows platform?

Reply:
For now I'm not worrying about Windows, since I am not familiar with that platform at all. Eventually, someone else will need to finish it. The good news is that there isn't that much that needs to be done.